File: packages/kotlin/src/main/kotlin/me/really/crypto/P384Ecdsa.kt
// SPDX-FileCopyrightText: Copyright © 2026 ReallyMe LLC. All rights reserved
//
// SPDX-License-Identifier: Apache-2.0

package me.really.crypto

import java.math.BigInteger
import java.security.MessageDigest
import java.security.SecureRandom
import org.bouncycastle.asn1.ASN1Integer
import org.bouncycastle.asn1.ASN1Sequence
import org.bouncycastle.asn1.DERSequence
import org.bouncycastle.asn1.sec.SECNamedCurves
import org.bouncycastle.crypto.digests.SHA384Digest
import org.bouncycastle.crypto.params.ECDomainParameters
import org.bouncycastle.crypto.params.ECPrivateKeyParameters
import org.bouncycastle.crypto.params.ECPublicKeyParameters
import org.bouncycastle.crypto.signers.ECDSASigner
import org.bouncycastle.crypto.signers.HMacDSAKCalculator

/**
 * Deterministic P-384 ECDSA over SHA-384 backed by BouncyCastle.
 *
 * The package contract needs reproducible DER signatures that match the Rust
 * vectors byte-for-byte, so this wrapper uses RFC 6979 nonces rather than JCA
 * provider randomness.
 */
public object ReallyMeP384Ecdsa {
    public const val SECRET_KEY_LENGTH: Int = 48
    public const val COMPRESSED_PUBLIC_KEY_LENGTH: Int = 49
    public const val UNCOMPRESSED_PUBLIC_KEY_LENGTH: Int = 97

    private val domain: ECDomainParameters =
        ECDomainParameters(SECNamedCurves.getByName("secp384r1"))

    public fun generateKeyPair(): Pair<ByteArray, ByteArray> {
        val random = SecureRandom()
        val secretKey = ByteArray(SECRET_KEY_LENGTH)
        repeat(1024) {
            random.nextBytes(secretKey)
            val scalar = BigInteger(1, secretKey)
            if (scalar.signum() > 0 && scalar < domain.n) {
                return Pair(derivePublicKey(secretKey), secretKey.copyOf())
            }
        }
        throw ReallyMeCryptoException.ProviderFailure()
    }

    public fun deriveKeyPair(secretKey: ByteArray): Pair<ByteArray, ByteArray> =
        Pair(derivePublicKey(secretKey), secretKey.copyOf())

    public fun derivePublicKey(secretKey: ByteArray): ByteArray {
        val scalar = validatedScalar(secretKey)
        return domain.g.multiply(scalar).normalize().getEncoded(true)
    }

    public fun sign(message: ByteArray, secretKey: ByteArray): ByteArray {
        val scalar = validatedScalar(secretKey)
        val digest = MessageDigest.getInstance("SHA-384").digest(message)
        val signer = ECDSASigner(HMacDSAKCalculator(SHA384Digest()))
        signer.init(true, ECPrivateKeyParameters(scalar, domain))
        val components = signer.generateSignature(digest)
        return encodeDerSignature(components[0], components[1])
    }

    public fun verify(signature: ByteArray, message: ByteArray, publicKey: ByteArray) {
        if (publicKey.size != COMPRESSED_PUBLIC_KEY_LENGTH &&
            publicKey.size != UNCOMPRESSED_PUBLIC_KEY_LENGTH
        ) {
            throw ReallyMeCryptoException.InvalidInput()
        }

        val point = try {
            domain.curve.decodePoint(publicKey)
        } catch (_: IllegalArgumentException) {
            throw ReallyMeCryptoException.InvalidInput()
        }
        val (r, s) = decodeDerSignature(signature)
        val digest = MessageDigest.getInstance("SHA-384").digest(message)
        val verifier = ECDSASigner()
        verifier.init(false, ECPublicKeyParameters(point, domain))
        if (!verifier.verifySignature(digest, r, s)) {
            throw ReallyMeCryptoException.InvalidSignature()
        }
    }

    private fun validatedScalar(secretKey: ByteArray): BigInteger {
        if (secretKey.size != SECRET_KEY_LENGTH) {
            throw ReallyMeCryptoException.InvalidInput()
        }
        val scalar = BigInteger(1, secretKey)
        if (scalar.signum() <= 0 || scalar >= domain.n) {
            throw ReallyMeCryptoException.InvalidInput()
        }
        return scalar
    }

    private fun encodeDerSignature(r: BigInteger, s: BigInteger): ByteArray =
        DERSequence(arrayOf(ASN1Integer(r), ASN1Integer(s))).encoded

    private fun decodeDerSignature(signature: ByteArray): Pair<BigInteger, BigInteger> {
        val sequence = try {
            ASN1Sequence.getInstance(signature)
        } catch (_: IllegalArgumentException) {
            throw ReallyMeCryptoException.InvalidInput()
        }
        if (sequence.size() != 2) {
            throw ReallyMeCryptoException.InvalidInput()
        }
        val r = ASN1Integer.getInstance(sequence.getObjectAt(0)).positiveValue
        val s = ASN1Integer.getInstance(sequence.getObjectAt(1)).positiveValue
        if (r.signum() <= 0 || r >= domain.n || s.signum() <= 0 || s >= domain.n) {
            throw ReallyMeCryptoException.InvalidInput()
        }
        return Pair(r, s)
    }
}

File: packages/ts/src/p384Ecdsa.ts
// SPDX-FileCopyrightText: Copyright © 2026 ReallyMe LLC. All rights reserved
//
// SPDX-License-Identifier: Apache-2.0

import { p384 } from "@noble/curves/nist.js";
import { sha384 } from "@noble/hashes/sha2.js";
import {
  decodeEcdsaDerSignature,
  encodeEcdsaDerSignature,
} from "./encodeEcdsaDer.js";
import { ReallyMeCryptoError } from "./errors.js";

export const P384_ECDSA_SECRET_KEY_LENGTH = 48;
export const P384_ECDSA_COMPRESSED_PUBLIC_KEY_LENGTH = 49;
export const P384_ECDSA_COMPACT_SIGNATURE_LENGTH = 96;
export const P384_ECDSA_DER_SIGNATURE_MAX_LENGTH = 104;

/**
 * P-384 ECDSA backed by @noble/curves.
 *
 * The package signs SHA-384(message) exactly once and preserves DER signatures
 * byte-for-byte with the Rust/Swift vector contract.
 */
export const ReallyMeP384Ecdsa = {
  generateKeyPair(): { publicKey: Uint8Array; secretKey: Uint8Array } {
    const secretKey = p384.utils.randomSecretKey();
    return {
      publicKey: p384.getPublicKey(secretKey, true),
      secretKey,
    };
  },

  deriveKeyPair(secretKey: Uint8Array): { publicKey: Uint8Array; secretKey: Uint8Array } {
    return {
      publicKey: this.derivePublicKey(secretKey),
      secretKey: secretKey.slice(),
    };
  },

  derivePublicKey(secretKey: Uint8Array): Uint8Array {
    if (secretKey.length !== P384_ECDSA_SECRET_KEY_LENGTH) {
      throw new ReallyMeCryptoError("invalid-input");
    }
    try {
      return p384.getPublicKey(secretKey, true);
    } catch {
      throw new ReallyMeCryptoError("invalid-input");
    }
  },

  sign(message: Uint8Array, secretKey: Uint8Array): Uint8Array {
    if (secretKey.length !== P384_ECDSA_SECRET_KEY_LENGTH) {
      throw new ReallyMeCryptoError("invalid-input");
    }
    try {
      const compactSignature = p384.sign(sha384(message), secretKey, {
        lowS: false,
        prehash: false,
      });
      return encodeEcdsaDerSignature(compactSignature, P384_ECDSA_SECRET_KEY_LENGTH);
    } catch {
      throw new ReallyMeCryptoError("invalid-input");
    }
  },

  verify(
    signature: Uint8Array,
    message: Uint8Array,
    publicKey: Uint8Array,
  ): void {
    if (publicKey.length !== P384_ECDSA_COMPRESSED_PUBLIC_KEY_LENGTH) {
      throw new ReallyMeCryptoError("invalid-input");
    }

    let compactSignature: Uint8Array;
    try {
      p384.Point.fromBytes(publicKey);
      compactSignature = decodeEcdsaDerSignature(
        signature,
        P384_ECDSA_SECRET_KEY_LENGTH,
        P384_ECDSA_DER_SIGNATURE_MAX_LENGTH,
      );
    } catch {
      throw new ReallyMeCryptoError("invalid-input");
    }

    try {
      const valid = p384.verify(compactSignature, sha384(message), publicKey, {
        lowS: false,
        prehash: false,
      });
      if (!valid) {
        throw new ReallyMeCryptoError("invalid-signature");
      }
    } catch {
      throw new ReallyMeCryptoError("invalid-signature");
    }
  },
} as const;
